scipy/stats/stats.py

```diff
@@ -8121,6 +8121,281 @@ def brunnermunzel(x, y, alternative="two-sided", distribution="t",
+    two one-sided confidence intervals, or the lower and upper bounds of a
+    two-sided confidence interval.
+
+    If `x` is the set of samples, the return values are the sample values
+    corresponding with the bounds of the confidence interval. If `x` is an
+    integer, the function considers `x` as the sample size and returns the
+    indexes of the bounds for the confidence interval for this sample size. Note
+    that this is possible because the numerical values of the data are
+    irrelevant for this computing confidence interval using this method.
+
+    A similar function is available in the QuantileNPCI R package [3]_. The
+    foundation is the same, but it computes the confidence interval bounds by
+    doing interpolations between the sample values, whereas this function uses
+    only sample values as bounds. Thus, `confint_quantile` returns slightly more
+    conservative intervals (i.e., larger).
+
+    Parameters
+    ----------
+    x : array_like or int Array of samples, should be one-dimensional. If
+        integer, taken as the number of samples available (strictly positive)
+    quantile : float The quantile for which we want to compute the confidence 
+        interval. Must be strictly between 0 and 1. 
+    confidence : float The desired confidence level of the confidence interval. 
+        Must be strictly between 0 and 1.
+    type : {'one-sided', 'two-sided'}, optional
+        Defines the type of confidence interval computed. Default is
+        'one-sided'.
+
+          * 'one-sided' :
+
+          computes the best possible one-sided confidence intervals
+          (both lower and upper bounds) for the given quantile.
+
+          * 'two-sided' :
+
+          computes a two-sided confidence interval by combination of
+          two one-sided intervals. E.g., a 90% two-sided interval is
+          computed by combining two 95% one-sided intervals.
+
+    Returns
+    -------
+    LB : float or int or `None` value or index of the lower bound of
+
+        * the right-open one-sided confidence interval (default),
+        * a two-sided confidence interval
+
+        `None` is returned when there are not enough samples to compute
+        the confidence interval with the desired level of confidence.
+    UB : float or int  or None value or index of the upper bound of
+
+        * the left-open one-sided confidence interval (default),
+        * a two-sided confidence interval
+
+        `None` is returned when there are not enough samples to compute
+        the confidence interval with the desired level of confidence.
+
+    Notes
+    -----
+    Two-sided confidence intervals are not guaranteed to be optimal. I.e., there
+    may exist a tighter interval that may contain the quantile of interest with
+    probability larger than the confidence level. These intervals may be found
+    by exhaustive search, which we do not do for efficiency reasons.
```

Comment:
Can you describe that? Or was this in an old version of the code?

Reply:
This was never part of the code in this PR. It is simply to explain (hopefully) clearly how the two-sided intervals are computed.

The point is that I chose here to construct two-sided intervals by concatenation of two one-sided ones, because it is a simple approach. But that's not necessarily the best CI that could exist. For example, let's say we make a two-sided 90% CI with two one-sided 95% ones. The bounds returned guarantee that they are the best (ie, most extreme) indexes that exclude **at least** 5% of the probability distribution below (for the lb) and above (for the ub) them. 

But in general they exclude more than 5% (even more so when the number of samples is small), and so our two-sided 90% may actually be a 92 or 93% CI. And in some cases, you could exclude one extra sample which would yield to a strictly tighter CI that still contains more that 90% of the probability distribution. 
...
Not sure if all that makes sense. Kinda hard without a whiteboard ^^'

Reply:
Thanks. I think I understand. 
More generally, are we trying to minimize over `i` and `j` `bd.cdf(j) - bd.cdf(i) - confidence_level` subject to the constraint that `bd.cdf(j) - bd.cdf(i) > confidence_level`?

Reply:
Intuitively, I would say that we aim to minimize `j-i` but in this case, I expect it would be (always?) the same thing as minimizing `bd.cdf(j) - bd.cdf(i) - confidence_level`.

The core of the "problem" is that there exists in general a set of Pareto-optimal (i.e., incomparably good) solutions: for example, `[i, j]` and `[i+1, j+1]`.

And finally, the way we compute the two-sided interval is not guaranteed to return one of these Pareto-optimal solutions. It will in most cases, but not always. 
I chose to implement it this way because 
(1) it is much more computationally efficient and 
(2) the suboptimality is fairly small in general.

Reply:
Thanks. Makes sense.